apache/trafficcontrol
# Rewrite /federations to Go - POST/PUT/DELETE

## What does this PR (Pull Request) do?
- [x] This PR fixes #3790 

Rewrites `/federations` to Go. This also adds API version 1.4 specific logic that
- Adds the messages normally returned as the "response" object as alerts (doesn't replace `response`, much as I'd like to)
- Allows POST request bodies that don't use the redundant top-level "federations" key (still backward-compatible; just gives clients the choice)

## Which Traffic Control components are affected by this PR?
- Documentation
- Traffic Control Client (Go)
- Traffic Ops

## What is the best way to verify this PR?
Run the unit tests and API/Go client integration tests. 

### Test Manually (CDN-in-a-Box)
1. Build and start CDN-in-a-Box
2. In Traffic Portal, go to <kbd>CDNs</kbd>&#8594;<kbd>CDN-in-a-Box</kbd>&#8594;<kbd>More</kbd>&#8594;<kbd>Manage Federations</kbd>
3. Click on the "+" button
4. Fill out the form - make sure to assign the user to `admin` (since I assume that's the user as whom you've logged in) and hit "Submit" to create a new Federation
5. Make a `POST` request to `/federations`, then a `PUT` then a `DELETE` (afaik there's no mechanism for this in Traffic Portal, so use `curl` or the `toaccess` scripts or something).

### Test Manually (Locally-run servers)
1. Build and start Traffic Ops
2. Assuming you already have a CDN, a Delivery Service, and a user as whom to sign in, make a new Federation using the `POST` method of `/cdns/{{your CDN name here}}/federations`.
3. Assign your federation to a Delivery Service (within your user's Tenant - I don't know if this will check) using the `POST` method of `/federations/{{ID of the federation you just created - probably 1}}/deliveryservices`
4. Assign the Federation to your user using the `POST` method of `/federations/{{same ID again}}/users`
5. Make a `POST` request to `/federations`, then a `PUT` then a `DELETE`

## The following criteria are ALL met by this PR
